File: partiql-plan/src/main/java/org/partiql/plan/rel/RelFilter.java
package org.partiql.plan.rel;

import org.jetbrains.annotations.NotNull;
import org.partiql.plan.Operand;
import org.partiql.plan.OperatorVisitor;
import org.partiql.plan.rex.Rex;

import java.util.List;

/**
 * Logical filter abstract base class.
 */
public abstract class RelFilter extends RelBase {

    /**
     * @return new {@link RelFilter} instance
     */
    @NotNull
    public static RelFilter create(@NotNull Rel input, @NotNull Rex predicate) {
        return new Impl(input, predicate);
    }

    /**
     * @return input rel (operand 0)
     */
    @NotNull
    public abstract Rel getInput();

    /**
     * @return predicate rex.
     */
    @NotNull
    public abstract Rex getPredicate();

    @NotNull
    @Override
    protected final RelType type() {
        throw new UnsupportedOperationException("Derive type is not implemented");
    }

    @NotNull
    @Override
    protected final List<Operand> operands() {
        Operand c0 = Operand.single(getInput());
        return List.of(c0);
    }

    @Override
    public <R, C> R accept(@NotNull OperatorVisitor<R, C> visitor, C ctx) {
        return visitor.visitFilter(this, ctx);
    }

    @NotNull
    public abstract RelFilter copy(@NotNull Rel input);

    @NotNull
    public abstract RelFilter copy(@NotNull Rel input, @NotNull Rex predicate);

    private static class Impl extends RelFilter {

        private final Rel input;
        private final Rex predicate;

        private Impl(Rel input, Rex predicate) {
            this.input = input;
            this.predicate = predicate;
        }

        @NotNull
        @Override
        public Rel getInput() {
            return input;
        }

        @NotNull
        @Override
        public Rex getPredicate() {
            return predicate;
        }

        @NotNull
        @Override
        public RelFilter copy(@NotNull Rel input) {
            return new Impl(input, predicate);
        }

        @NotNull
        @Override
        public RelFilter copy(@NotNull Rel input, @NotNull Rex predicate) {
            return new Impl(input, predicate);
        }
    }
}

File: partiql-plan/src/main/kotlin/org/partiql/plan/rel/RelFilter.kt
package org.partiql.plan.rel

import org.partiql.plan.Visitor
import org.partiql.plan.rex.Rex

/**
 * Logical filter operation for the WHERE and HAVING clauses.
 *
 *   arg 0 – rel input
 *   arg 1 - rex predicate
 */
public interface RelFilter : Rel {

    public fun getInput(): Rel

    public fun getPredicate(): Rex

    override fun getChildren(): Collection<Rel> = listOf(getInput())

    override fun getType(): RelType = getInput().getType()

    override fun isOrdered(): Boolean = getInput().isOrdered()

    override fun <R, C> accept(visitor: Visitor<R, C>, ctx: C): R =
        visitor.visitFilter(this, ctx)
}

/**
 * Default [RelFilter] implementation.
 */
internal class RelFilterImpl(input: Rel, predicate: Rex) : RelFilter {

    // DO NOT USE FINAL
    private var _input: Rel = input
    private var _children: List<Rel>? = null
    private var _predicate: Rex = predicate
    private var _ordered: Boolean = input.isOrdered()

    override fun getInput(): Rel = _input

    override fun getChildren(): Collection<Rel> {
        if (_children == null) {
            _children = listOf(_input)
        }
        return _children!!
    }

    override fun getPredicate(): Rex = _predicate

    override fun getType(): RelType = _input.getType()

    override fun isOrdered(): Boolean = _ordered

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is RelFilter) return false
        if (_input != other.getInput()) return false
        if (_predicate != other.getPredicate()) return false
        return true
    }

    override fun hashCode(): Int {
        var result = 1
        result = 31 * result + _input.hashCode()
        result = 31 * result + _predicate.hashCode()
        return result
    }
}
